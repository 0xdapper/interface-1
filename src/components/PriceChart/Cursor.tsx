import { impactAsync, ImpactFeedbackStyle } from 'expo-haptics'
import React from 'react'
import { StyleSheet } from 'react-native'
import { Gesture, GestureDetector } from 'react-native-gesture-handler'
import {
  runOnJS,
  SharedValue,
  useAnimatedStyle,
  withSpring,
  withTiming,
} from 'react-native-reanimated'
import { getYForX, round } from 'react-native-redash'
import { useAppTheme } from 'src/app/hooks'
import { AnimatedBox, Box } from 'src/components/layout/Box'
import {
  AnimatedNumber,
  AnimatedTranslation,
  GraphMetadatas,
} from 'src/components/PriceChart/types'
import { CHART_HEIGHT } from 'src/components/PriceChart/utils'
import { opacify } from 'src/utils/colors'

const CURSOR_INNER_SIZE = 12
const CURSOR_SIZE = CURSOR_INNER_SIZE + 6
const LINE_WIDTH = 1

// Length in screen unit of a single tick
// TODO(MOB-2255): correlate ticks to scale
const TICK_LENGTH = 10

interface CursorProps {
  graphs: GraphMetadatas
  index: AnimatedNumber
  isActive: SharedValue<boolean>
  translation: AnimatedTranslation
  cursorColor?: NullUndefined<string>
}

export const Cursor = ({ graphs, index, isActive, translation, cursorColor }: CursorProps) => {
  const theme = useAppTheme()

  const panGesture = Gesture.Pan()
    .activeOffsetX([-10, 10])
    .onStart(() => {
      isActive.value = true
      runOnJS(impactAsync)(ImpactFeedbackStyle.Light)
    })
    .onUpdate((event) => {
      // First verify if we're crossing a tick for haptic feedback
      const hasCrossedTick =
        round(event.x / TICK_LENGTH) * TICK_LENGTH !==
        round(translation.x.value / TICK_LENGTH) * TICK_LENGTH
      if (hasCrossedTick) {
        runOnJS(impactAsync)(ImpactFeedbackStyle.Light)
      }

      // Then update value continuously
      translation.x.value = event.x
      const graph = graphs[index.value]
      if (graph) {
        translation.y.value = getYForX(graph.data.path, translation.x.value) || 0
      }
    })
    .onFinalize(() => (isActive.value = false))

  const containerStyles = useAnimatedStyle(() => ({
    opacity: withTiming(isActive.value ? 1 : 0),
  }))

  // cursor position and scale animation
  const cursorAnimatedStyles = useAnimatedStyle(() => {
    const translateX = translation.x.value - CURSOR_SIZE / 2
    const translateY = translation.y.value - CURSOR_SIZE / 2
    return {
      transform: [{ translateX }, { translateY }, { scale: withSpring(isActive.value ? 1 : 0) }],
    }
  })
  const verticalLineAnimatedStyles = useAnimatedStyle(() => {
    const translateX = translation.x.value - LINE_WIDTH
    return {
      transform: [{ translateX }, { scale: withTiming(isActive.value ? 1 : 0) }],
    }
  })

  return (
    <Box flex={1} style={StyleSheet.absoluteFill}>
      <GestureDetector gesture={panGesture}>
        <AnimatedBox style={[StyleSheet.absoluteFill, containerStyles]}>
          {/* Vertical line */}
          <AnimatedBox
            borderRadius="xs"
            borderStyle="dashed"
            borderWidth={1}
            height={CHART_HEIGHT}
            style={[
              verticalLineAnimatedStyles,
              StyleSheet.absoluteFill,
              {
                borderColor: cursorColor
                  ? opacify(50, cursorColor)
                  : theme.colors.backgroundOutline,
              },
            ]}
            width={LINE_WIDTH}
          />

          {/* Circular cursor */}
          <AnimatedBox
            alignItems="center"
            height={CURSOR_SIZE}
            justifyContent="center"
            position="relative"
            style={cursorAnimatedStyles}
            width={CURSOR_SIZE}>
            <Box
              borderRadius="full"
              height={CURSOR_INNER_SIZE}
              style={{ backgroundColor: cursorColor ?? theme.colors.textSecondary }}
              width={CURSOR_INNER_SIZE}
            />
          </AnimatedBox>
        </AnimatedBox>
      </GestureDetector>
    </Box>
  )
}
