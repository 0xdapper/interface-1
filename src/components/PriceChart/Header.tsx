import { createRestyleComponent, createVariant, useTheme, VariantProps } from '@shopify/restyle'
import React from 'react'
import { interpolate, useAnimatedStyle, useDerivedValue } from 'react-native-reanimated'
import { ReText, round } from 'react-native-redash'
import { Box } from 'src/components/layout/Box'
import { HEIGHT } from 'src/components/PriceChart/Model'
import { AnimatedIndex, AnimatedTranslation, GraphMetadatas } from 'src/components/PriceChart/types'
import { Theme } from 'src/styles/theme'

interface HeaderProps {
  translation: AnimatedTranslation
  index: AnimatedIndex
  graphs: GraphMetadatas
}

// const StyledReText = createBox<Theme,  & typeof ReText.arguments>(ReText)

const StyledReText = createRestyleComponent<
  VariantProps<Theme, 'textVariants'> & React.ComponentProps<typeof ReText>,
  Theme
>([createVariant({ themeKey: 'textVariants' })], ReText)

export const Header = ({ translation, index, graphs }: HeaderProps) => {
  const theme = useTheme<Theme>()

  const data = useDerivedValue(() => graphs[index.value].data)

  const price = useDerivedValue(() =>
    interpolate(translation.y.value, [0, HEIGHT], [data.value.maxPrice, data.value.minPrice])
  )
  const percentChange = useDerivedValue(
    () => ((price.value - data.value.startingPrice) / data.value.startingPrice) * 100
  )

  const priceFormatted = useDerivedValue(() => {
    return `$${round(price.value, 2).toLocaleString('en-US', { currency: 'USD' })}`
  })
  const percentChangeFormatted = useDerivedValue(() => `${round(percentChange.value, 3)}%`)

  const style = useAnimatedStyle(() => ({
    color: percentChange.value > 0 ? theme.colors.green : theme.colors.red,
  }))

  return (
    <Box flex={1} padding="md" justifyContent="center">
      <Box flex={1} alignItems="center">
        <StyledReText variant="h1" text={priceFormatted} />
        <StyledReText variant="h3" style={style} text={percentChangeFormatted} />
      </Box>
    </Box>
  )
}
