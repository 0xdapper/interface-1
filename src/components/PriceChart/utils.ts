import { scaleLinear } from 'd3-scale'
import { curveBasis, line } from 'd3-shape'
import { getYForX, parse, Path } from 'react-native-redash'
import { GraphData, Price, PriceList } from 'src/components/PriceChart/types'
import { dimensions } from 'src/styles/sizing'
import { theme as FixedTheme } from 'src/styles/theme'
import { logger } from 'src/utils/logger'

// sets the height of the chart short enough on small devices that the
// "Your balance" section will always show above the fold
// we can't use useResponsiveProps for this because CHART_HEIGHT gets
// used in non-component code related to chart functionality
export const CHART_HEIGHT = dimensions.fullHeight < FixedTheme.breakpoints.sm.height ? 180 : 310
export const CHART_WIDTH = dimensions.fullWidth

export const NUM_GRAPHS = 5
// number of points in graph
// will interpolate up or down to the precision
// 365 is a balance between performance and precision
// most paths have fewer than or exactly 365 points
export const GRAPH_PRECISION = 365

/**
 * Normalizes a Path by rescaling the x-axis and mapping to the Path.
 * Required because different date ranges have different number of points.
 * redash requires the same number of points to interpolate between paths.
 * @param path to normalize
 * @param precision number of dat points in the new path
 * @param width of the canvas
 **/
export function normalizePath(path: Path, precision: number, width: number): Path {
  const scaleX = scaleLinear()
    .domain([0, precision - 1])
    .range([0, width])

  const values = Array(precision)
    .fill([])
    .map((_, index) => {
      const x = scaleX(index)
      const y = getYForX(path, x)
      return [x, y] as [number, number]
    })

  return parse(
    line()
      .x(([x]) => x)
      .y(([, y]) => y)
      .curve(curveBasis)(values) as string
  )
}

/** Constructs a drawable Path from a PriceList */
export function buildGraph(
  priceList: NullUndefined<PriceList>,
  precision: number,
  width = CHART_WIDTH,
  height = CHART_HEIGHT
): GraphData | null {
  if (!priceList || priceList.length === 0) return null
  priceList = priceList.slice().reverse()

  const formattedValues = priceList
    .filter((value): value is Price => !!value)
    .map((price) => [price.timestamp, price.close] as [number, number])

  const prices = formattedValues.map(([, price]) => price)
  const dates = formattedValues.map(([date]) => date)

  const lowPrice = Math.min(...prices)
  const highPrice = Math.max(...prices)
  const openPrice = prices[0]
  const closePrice = prices[prices.length - 1]

  const openDate = dates[0]
  const closeDate = dates[dates.length - 1]

  if (
    openDate === undefined ||
    closeDate === undefined ||
    openPrice === undefined ||
    closePrice === undefined
  )
    return null

  // TODO: [MOB-3876] consider using `scaleTime`
  const scaleX = scaleLinear().domain([openDate, closeDate]).range([0, width])
  const scaleY = scaleLinear().domain([lowPrice, highPrice]).range([height, 0])

  try {
    // normalize brings all paths to the same precision (number of points)
    const path = normalizePath(
      parse(
        line()
          .x(([x]) => scaleX(x))
          .y(([, y]) => scaleY(y))
          .curve(curveBasis)(formattedValues) as string
      ),
      precision,
      width
    )

    return {
      openDate,
      closeDate,
      lowPrice,
      highPrice,
      openPrice,
      closePrice,
      path,
    }
  } catch (e: unknown) {
    logger.error(
      'PriceChart/utils',
      'buildGraph',
      `Error while normalizing path. ${formattedValues}`
    )
    return null
  }
}
