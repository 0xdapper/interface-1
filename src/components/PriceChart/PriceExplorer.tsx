import { useTheme } from '@shopify/restyle'
import React from 'react'
import { StyleSheet, TouchableWithoutFeedback, View } from 'react-native'
import Animated, {
  useAnimatedProps,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated'
import { mixPath, useVector } from 'react-native-redash'
import Svg, { Defs, LinearGradient, Path, Stop } from 'react-native-svg'
import { AnimatedBox, Box } from 'src/components/layout/Box'
import { Cursor } from 'src/components/PriceChart/Cursor'
import { Header } from 'src/components/PriceChart/Header'
import { HEIGHT, NUM_GRAPHS, WIDTH } from 'src/components/PriceChart/Model'
import { TimeRangeLabel } from 'src/components/PriceChart/TimeRangeLabel'
import { GraphMetadatas } from 'src/components/PriceChart/types'
import { Theme } from 'src/styles/theme'

const AnimatedPath = Animated.createAnimatedComponent(Path)

const SELECTION_WIDTH = WIDTH - 50
const BUTTON_WIDTH = SELECTION_WIDTH / NUM_GRAPHS

interface GraphProps {
  graphs: GraphMetadatas
  title: string
}

/**
 * Displays a set of graphs with header, scrubbable chart and navigation row.
 * Inspired by https://github.com/wcandillon/can-it-be-done-in-react-native/tree/master/season4/src/Rainbow
 */
export const PriceExplorer = ({ graphs, title }: GraphProps) => {
  const theme = useTheme<Theme>()

  // whether the graph pan gesture is active
  const isPanning = useSharedValue(false)
  // pan gesture x and y
  const translation = useVector()
  // used in animations
  const transition = useSharedValue(1)
  // previous graph index used in animations
  const previousGraphIndex = useSharedValue<number>(0)
  // graph index used to display the current graph
  const currentGraphIndex = useSharedValue<number>(0)

  // mixes graph paths on index change
  const graphTransitionAnimatedProps = useAnimatedProps(() => {
    const previousPath = graphs[previousGraphIndex.value].data.path
    const currentPath = graphs[currentGraphIndex.value].data.path
    const d = {
      d: mixPath(transition.value, previousPath, currentPath),
    }
    return d
  })
  const graphTransitionClosedAnimatedProps = useAnimatedProps(() => {
    const previousPath = graphs[previousGraphIndex.value].data.path
    const currentPath = graphs[currentGraphIndex.value].data.path
    return {
      d: `${mixPath(
        transition.value,
        previousPath,
        currentPath
      )} L ${WIDTH} ${HEIGHT} L 0 ${HEIGHT}`,
    }
  })

  // animates slider (time range label background) on press
  const sliderStyle = useAnimatedStyle(() => ({
    transform: [{ translateX: withTiming(BUTTON_WIDTH * currentGraphIndex.value) }],
  }))

  return (
    <Box flex={1}>
      <Header
        graphs={graphs}
        index={currentGraphIndex}
        isPanning={isPanning}
        title={title}
        translation={translation}
      />
      <Box mb="sm">
        <Svg height={HEIGHT} width={WIDTH}>
          <AnimatedPath
            animatedProps={graphTransitionAnimatedProps}
            fill="transparent"
            stroke={theme.colors.primary1}
            strokeWidth={2}
          />

          <Defs>
            <LinearGradient id="gradient" x1="50%" x2="50%" y1="0%" y2="100%">
              <Stop offset="0%" stopColor={theme.colors.primary1} stopOpacity="0.2" />
              <Stop offset="100%" stopColor={theme.colors.primary1} stopOpacity="0" />
            </LinearGradient>
          </Defs>

          <AnimatedPath
            animatedProps={graphTransitionClosedAnimatedProps}
            fill="url(#gradient)"
            stroke="transparent"
            strokeWidth={3}
          />
        </Svg>
        <Cursor
          graphs={graphs}
          index={currentGraphIndex}
          isActive={isPanning}
          translation={translation}
        />
      </Box>
      <Box alignSelf="center" flexDirection="row" width={SELECTION_WIDTH}>
        <View style={StyleSheet.absoluteFill}>
          <AnimatedBox
            bg="primary1"
            borderRadius="lg"
            style={[StyleSheet.absoluteFillObject, sliderStyle]}
            width={BUTTON_WIDTH}
          />
        </View>
        {graphs.map((graph, index) => {
          return (
            <TouchableWithoutFeedback
              key={graph.label}
              onPress={() => {
                previousGraphIndex.value = currentGraphIndex.value
                transition.value = 0
                currentGraphIndex.value = index
                transition.value = withTiming(1)
              }}>
              <Box padding="xs" width={BUTTON_WIDTH}>
                <TimeRangeLabel
                  index={index}
                  label={graph.label}
                  selectedIndex={currentGraphIndex}
                  transition={transition}
                />
              </Box>
            </TouchableWithoutFeedback>
          )
        })}
      </Box>
    </Box>
  )
}
