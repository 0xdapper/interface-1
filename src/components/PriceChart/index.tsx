import { Currency } from '@uniswap/sdk-core'
import React, { ReactNode, Suspense, useMemo } from 'react'
import { useTranslation } from 'react-i18next'
import { Flex } from 'src/components/layout'
import { Box } from 'src/components/layout/Box'
import { PriceChartLoading } from 'src/components/PriceChart/PriceChartLoading'
import { PriceExplorer } from 'src/components/PriceChart/PriceExplorer'
import { useTokenPriceGraphs } from 'src/components/PriceChart/TokenModel'
import { GraphMetadatas } from 'src/components/PriceChart/types'
import { Text } from 'src/components/Text'

export const CurrencyPriceChart = ({ currency }: { currency: Currency }) => {
  return (
    <Suspense fallback={<PriceChartLoading />}>
      <CurrencyPriceChartInner currency={currency} />
    </Suspense>
  )
}

function CurrencyPriceChartInner({ currency }: { currency: Currency }) {
  const graphs = useTokenPriceGraphs(currency.wrapped)
  const { t } = useTranslation()

  const error = graphs === null

  if (error) {
    return (
      <Flex centered mx="lg" my="md">
        <Text color="accentCritical" textAlign="center" variant="bodyLarge">
          {t('Could not retrieve price history')}
        </Text>
      </Flex>
    )
  }

  return <PriceChart graphs={graphs} />
}

export function PriceChart({
  graphs,
  customChartLabel,
}: {
  graphs?: GraphMetadatas
  customChartLabel?: ReactNode
}) {
  // require all graphs to be loaded before rendering the chart
  // TODO(judo): improve loading state
  const loading =
    useMemo(() => graphs?.some((g) => g.data === undefined), [graphs]) || graphs === undefined

  return (
    <Box overflow="hidden">
      {loading ? (
        <PriceChartLoading />
      ) : (
        <PriceExplorer customChartLabel={customChartLabel} graphs={graphs} />
      )}
    </Box>
  )
}
