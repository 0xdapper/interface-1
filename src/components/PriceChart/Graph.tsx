import React from 'react'
import { StyleSheet, TouchableWithoutFeedback, View } from 'react-native'
import Animated, {
  useAnimatedProps,
  useAnimatedStyle,
  useSharedValue,
  withTiming,
} from 'react-native-reanimated'
import { mixPath, useVector } from 'react-native-redash'
import Svg, { Defs, LinearGradient, Path, Stop } from 'react-native-svg'
import { AnimatedBox, Box } from 'src/components/layout/Box'
import { Cursor } from 'src/components/PriceChart/Cursor'
import { Header } from 'src/components/PriceChart/Header'
import { HEIGHT, NUM_GRAPHS, WIDTH } from 'src/components/PriceChart/Model'
import { GraphMetadatas } from 'src/components/PriceChart/types'
import { Text } from 'src/components/Text'

const AnimatedPath = Animated.createAnimatedComponent(Path)

const SELECTION_WIDTH = WIDTH - 50
const BUTTON_WIDTH = SELECTION_WIDTH / NUM_GRAPHS

interface GraphProps {
  graphs: GraphMetadatas
}

/**
 * Displays a set of graphs with header, scrubbable chart and navigation row.
 * Inspired by https://github.com/wcandillon/can-it-be-done-in-react-native/tree/master/season4/src/Rainbow
 */
export const Graph = ({ graphs }: GraphProps) => {
  const translation = useVector()
  const transition = useSharedValue(0)

  const previousGraphIndex = useSharedValue<number>(0)
  const currentGraphIndex = useSharedValue<number>(0)

  // mixes graph paths on index change
  const graphTransitionAnimatedProps = useAnimatedProps(() => {
    const previousPath = graphs[previousGraphIndex.value].data.path
    const currentPath = graphs[currentGraphIndex.value].data.path
    return {
      //TODO(judo): make animation interruptible
      d: mixPath(transition.value, previousPath, currentPath),
    }
  })
  const graphTransitionClosedAnimatedProps = useAnimatedProps(() => {
    //TODO(judo): figure out a way to merge with `graphTransitionAnimatedProps`
    const previousPath = graphs[previousGraphIndex.value].data.path
    const currentPath = graphs[currentGraphIndex.value].data.path
    return {
      d: `${mixPath(
        transition.value,
        previousPath,
        currentPath
      )} L ${WIDTH} ${HEIGHT} L 0 ${HEIGHT}`,
    }
  })

  const sliderStyle = useAnimatedStyle(() => ({
    transform: [{ translateX: withTiming(BUTTON_WIDTH * currentGraphIndex.value) }],
  }))

  return (
    <Box flex={1}>
      <Header translation={translation} index={currentGraphIndex} graphs={graphs} />
      <View>
        <Svg width={WIDTH} height={HEIGHT}>
          <AnimatedPath
            animatedProps={graphTransitionAnimatedProps}
            fill="transparent"
            stroke="#FF007A"
            strokeWidth={2}
          />

          <Defs>
            <LinearGradient x1="50%" y1="0%" x2="50%" y2="100%" id="gradient">
              <Stop stopColor="#FF007A" offset="0%" stopOpacity="0.2" />
              <Stop stopColor="#FF007A" offset="100%" stopOpacity="0" />
            </LinearGradient>
          </Defs>

          <AnimatedPath
            animatedProps={graphTransitionClosedAnimatedProps}
            fill="url(#gradient)"
            stroke="transparent"
            strokeWidth={3}
          />
        </Svg>
        <Cursor translation={translation} index={currentGraphIndex} graphs={graphs} />
      </View>
      <Box flexDirection="row" width={SELECTION_WIDTH} alignSelf="center">
        <View style={StyleSheet.absoluteFill}>
          <AnimatedBox
            bg="pink"
            borderRadius="lg"
            width={BUTTON_WIDTH}
            style={[StyleSheet.absoluteFillObject, sliderStyle]}
          />
        </View>
        {graphs.map((graph, index) => {
          return (
            <TouchableWithoutFeedback
              key={graph.label}
              onPress={() => {
                previousGraphIndex.value = currentGraphIndex.value
                transition.value = 0
                currentGraphIndex.value = index
                transition.value = withTiming(1)
              }}>
              <AnimatedBox padding="xs" width={BUTTON_WIDTH}>
                <Text
                  variant="buttonLabel"
                  color={index === currentGraphIndex.value ? 'white' : 'primary1'}
                  textAlign="center">
                  {graph.label}
                </Text>
              </AnimatedBox>
            </TouchableWithoutFeedback>
          )
        })}
      </Box>
    </Box>
  )
}
