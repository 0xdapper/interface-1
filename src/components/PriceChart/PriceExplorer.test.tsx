import React from 'react'
import 'react-native'
import 'react-native-gesture-handler'
import { PriceExplorer } from 'src/components/PriceChart/PriceExplorer'
import { buildGraph } from 'src/components/PriceChart/utils'
import { renderWithTheme } from 'src/test/render'

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
const graphDatapoints = buildGraph(
  [
    {
      timestamp: 0,
      close: 1551331.3403,
    },
    {
      timestamp: 1,
      close: 3102662.6806,
    },
  ],
  3
)!

const buildGraphMetadata = (index: number) =>
  ({
    label: `Graph ${index}`,
    index,
    data: graphDatapoints,
  } as const)

const graphs = [
  buildGraphMetadata(0),
  buildGraphMetadata(1),
  buildGraphMetadata(2),
  buildGraphMetadata(3),
  buildGraphMetadata(4),
] as const

it('renders correctly', () => {
  const tree = renderWithTheme(<PriceExplorer graphs={graphs} />)
  expect(tree).toMatchSnapshot()
})
