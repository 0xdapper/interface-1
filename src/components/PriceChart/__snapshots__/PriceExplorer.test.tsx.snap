// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "display": "flex",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "marginBottom": 24,
          "marginHorizontal": 16,
        },
      ]
    }
  >
    <TextInput
      animatedStyle={
        Object {
          "value": Object {},
        }
      }
      collapsable={false}
      editable={false}
      style={
        Object {
          "color": "#070A15",
          "fontFamily": "Inter-Medium",
          "fontSize": 36,
          "fontWeight": "300",
          "lineHeight": 44,
        }
      }
      text="$100"
      underlineColorAndroid="transparent"
      value="$100"
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "stretch",
            "display": "flex",
            "flexDirection": "row",
            "justifyContent": "flex-start",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "stretch",
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "flex-start",
            },
          ]
        }
      >
        <TextInput
          animatedStyle={
            Object {
              "value": Object {},
            }
          }
          collapsable={false}
          editable={false}
          style={
            Object {
              "color": "#5F667D",
              "fontFamily": "Inter-Regular",
              "fontSize": 12,
              "lineHeight": 16,
            }
          }
          text="-50%"
          underlineColorAndroid="transparent"
          value="-50%"
        />
        <View
          style={
            Array [
              Object {
                "display": "flex",
                "flexGrow": 0,
                "flexShrink": 0,
                "height": 0,
                "width": 0,
              },
            ]
          }
        />
        <TextInput
          animatedStyle={
            Object {
              "value": Object {
                "color": "#FD4040",
              },
            }
          }
          collapsable={false}
          editable={false}
          style={
            Object {
              "color": "#FD4040",
              "fontFamily": "Inter-Regular",
              "fontSize": 12,
              "lineHeight": 16,
            }
          }
          text="↘"
          underlineColorAndroid="transparent"
          value="↘"
        />
      </View>
      <View
        style={
          Array [
            Object {
              "display": "flex",
              "flexGrow": 0,
              "flexShrink": 0,
              "height": 4,
              "width": 4,
            },
          ]
        }
      />
      <TextInput
        animatedStyle={
          Object {
            "value": Object {},
          }
        }
        collapsable={false}
        editable={false}
        style={
          Object {
            "color": "#5F667D",
            "fontFamily": "Inter-Regular",
            "fontSize": 12,
            "lineHeight": 16,
          }
        }
        text=""
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "marginVertical": 12,
        },
      ]
    }
  >
    <RNSVGSvgView
      bbHeight={180}
      bbWidth={750}
      focusable={false}
      height={180}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          Object {
            "flex": 0,
            "height": 180,
            "width": 750,
          },
        ]
      }
      width={750}
    >
      <RNSVGGroup>
        <RNSVGPath
          d="M0,0 C0,0 62.5,15 62.5,15 C125,30 250,60 375,90 C500,120 625,150 687.5,165 C687.5,165 750,180 750,180"
          fill={0}
          propList={
            Array [
              "fill",
              "stroke",
              "strokeWidth",
            ]
          }
          stroke={4293867150}
          strokeWidth={2}
        />
        <RNSVGDefs>
          <RNSVGLinearGradient
            gradient={
              Array [
                0,
                871315086,
                1,
                15677070,
              ]
            }
            gradientTransform={null}
            gradientUnits={0}
            name="gradient"
            x1="50%"
            x2="50%"
            y1="0%"
            y2="100%"
          />
        </RNSVGDefs>
        <RNSVGPath
          d="M0,0 C0,0 62.5,15 62.5,15 C125,30 250,60 375,90 C500,120 625,150 687.5,165 C687.5,165 750,180 750,180 L 750 180 L 0 180"
          fill={
            Array [
              1,
              "gradient",
            ]
          }
          propList={
            Array [
              "fill",
              "stroke",
              "strokeWidth",
            ]
          }
          stroke={0}
          strokeWidth={3}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "flex": 1,
          },
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
        ]
      }
    >
      <View
        animatedStyle={
          Object {
            "value": Object {},
          }
        }
        collapsable={false}
        forwardedRef={[Function]}
        handlerTag={1}
        handlerType="PanGestureHandler"
        onGestureHandlerEvent={[Function]}
        style={
          Array [
            Object {
              "display": "flex",
            },
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
          ]
        }
      >
        <View
          animatedStyle={
            Object {
              "value": Object {
                "transform": Array [
                  Object {
                    "translateX": -25,
                  },
                  Object {
                    "translateY": -25,
                  },
                  Object {
                    "scale": 0,
                  },
                ],
              },
            }
          }
          collapsable={false}
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderRadius": 999999,
                "display": "flex",
                "height": 50,
                "justifyContent": "center",
                "width": 50,
              },
              Object {
                "backgroundColor": "rgba(243, 71, 191, 0.171)",
                "transform": Array [
                  Object {
                    "translateX": -25,
                  },
                  Object {
                    "translateY": -25,
                  },
                  Object {
                    "scale": 0,
                  },
                ],
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#EF368E",
                  "borderRadius": 999999,
                  "display": "flex",
                  "height": 15,
                  "width": 15,
                },
              ]
            }
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignSelf": "center",
          "display": "flex",
          "flexDirection": "row",
          "width": 750,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        animatedStyle={
          Object {
            "value": Object {
              "transform": Array [
                Object {
                  "translateX": 150,
                },
              ],
            },
          }
        }
        collapsable={false}
        style={
          Array [
            Object {
              "backgroundColor": "#F4F6FE",
              "borderRadius": 16,
              "display": "flex",
              "width": 150,
            },
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "translateX": 150,
                },
              ],
            },
          ]
        }
      />
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "display": "flex",
            "padding": 8,
            "width": 150,
          },
        ]
      }
    >
      <Text
        animatedStyle={
          Object {
            "value": Object {
              "color": "#5F667D",
            },
          }
        }
        collapsable={false}
        style={
          Object {
            "color": "#5F667D",
            "display": "flex",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 14,
            "lineHeight": 16,
            "textAlign": "center",
          }
        }
      >
        Graph 0
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "display": "flex",
            "padding": 8,
            "width": 150,
          },
        ]
      }
    >
      <Text
        animatedStyle={
          Object {
            "value": Object {
              "color": "rgba(7, 10, 21, 1)",
            },
          }
        }
        collapsable={false}
        style={
          Object {
            "color": "rgba(7, 10, 21, 1)",
            "display": "flex",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 14,
            "lineHeight": 16,
            "textAlign": "center",
          }
        }
      >
        Graph 1
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "display": "flex",
            "padding": 8,
            "width": 150,
          },
        ]
      }
    >
      <Text
        animatedStyle={
          Object {
            "value": Object {
              "color": "#5F667D",
            },
          }
        }
        collapsable={false}
        style={
          Object {
            "color": "#5F667D",
            "display": "flex",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 14,
            "lineHeight": 16,
            "textAlign": "center",
          }
        }
      >
        Graph 2
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "display": "flex",
            "padding": 8,
            "width": 150,
          },
        ]
      }
    >
      <Text
        animatedStyle={
          Object {
            "value": Object {
              "color": "#5F667D",
            },
          }
        }
        collapsable={false}
        style={
          Object {
            "color": "#5F667D",
            "display": "flex",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 14,
            "lineHeight": 16,
            "textAlign": "center",
          }
        }
      >
        Graph 3
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "display": "flex",
            "padding": 8,
            "width": 150,
          },
        ]
      }
    >
      <Text
        animatedStyle={
          Object {
            "value": Object {
              "color": "#5F667D",
            },
          }
        }
        collapsable={false}
        style={
          Object {
            "color": "#5F667D",
            "display": "flex",
            "fontFamily": "Inter-SemiBold",
            "fontSize": 14,
            "lineHeight": 16,
            "textAlign": "center",
          }
        }
      >
        Graph 4
      </Text>
    </View>
  </View>
</View>
`;
